protocols/mp-ecdsa/src/client.rs

```diff
@@ -0,0 +1,241 @@
+where
+    B: Block,
+    BE: Backend<B>,
+    T: BlockchainEvents<B> + HeaderBackend<B> + ProvideRuntimeApi<B> + Send + Sync,
+    <T as ProvideRuntimeApi<B>>::Api:
+        DKGApi<B, AuthorityId, NumberFor<B>, MaxProposalLength, MaxAuthorities>,
+{
+}
+
+impl<B: Block, BE: Backend<B>, C: ClientWithApi<B, BE>> MpEcdsaClient<B, BE, C> {
+    pub async fn validator_set(
+        &self,
+        header: &B::Header,
+    ) -> Option<(
+        AuthoritySet<AuthorityId, MaxAuthorities>,
+        AuthoritySet<AuthorityId, MaxAuthorities>,
+    )> {
+        let new = if let Some((new, queued)) = find_authorities_change::<B>(header) {
+            Some((new, queued))
+        } else {
+            let at = header.hash();
+            let current_authority_set = exec_client_function(&self.client, move |client| {
+                client.runtime_api().authority_set(at).ok()
+            })
+            .await;
+
+            let queued_authority_set = exec_client_function(&self.client, move |client| {
+                client.runtime_api().queued_authority_set(at).ok()
+            })
+            .await;
+
+            match (current_authority_set, queued_authority_set) {
+                (Some(current), Some(queued)) => Some((current, queued)),
+                _ => None,
+            }
+        };
+
+        self.logger
+            .trace(format!("🕸️  active validator set: {new:?}"));
+
+        new
+    }
+
+    pub async fn should_execute_new_keygen(
+        &self,
+        header: &B::Header,
+    ) -> AnticipatedKeygenExecutionStatus {
+        // query runtime api to check if we should execute new keygen.
+        let at = header.hash();
+        let (execute, force_execute) = self
+            .exec_client_function(move |client| {
+                client.runtime_api().should_execute_new_keygen(at).unwrap_or_default()
+            })
+            .await;
+
+        AnticipatedKeygenExecutionStatus { execute, force_execute }
+    }
+
+    /// Get the next DKG public key
+    pub async fn get_next_dkg_pub_key(
+        &self,
+        header: &B::Header,
+    ) -> Option<(AuthoritySetId, Vec<u8>)> {
+        let at = header.hash();
+        self.exec_client_function(move |client| {
+            client.runtime_api().next_dkg_pub_key(at).unwrap_or_default()
+        })
+            .await
+    }
+
+    pub async fn dkg_pub_key_is_unset(&self, header: &B::Header) -> bool {
+        self.get_dkg_pub_key(header).await.1.is_empty()
+    }
+
+    /// Get the active DKG public key
+    pub async fn get_dkg_pub_key(&self, header: &B::Header) -> (AuthoritySetId, Vec<u8>) {
+        let at = header.hash();
+        self.exec_client_function(move |client| {
+            client.runtime_api().dkg_pub_key(at).unwrap_or_default()
+        })
+            .await
+    }
+
+    /// Get the party index of our worker
+    ///
+    /// Returns `None` if we are not in the best authority set
+    pub async fn get_party_index(&self, header: &B::Header) -> Option<u16> {
```

Comment:
There will be no best authorities on Tangle Mainnet as is currently implemented by `dkg-substrate`. There is just the job and the participants of the job. The party index for a validator is their index in the job's participants.